### PR TITLE
mon: Remove any pg_upmap_primary mapping during remove a pool

### PR DESCRIPTION
**Pull request by mohit84**
The pg_upmap_primary items are not deleted after the pool is deleted from the cluster.

Fixes: https://tracker.ceph.com/issues/66867

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by mohit84** [comment]
jenkins test make check

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -14629,6 +14629,15 @@ int OSDMonitor::_prepare_remove_pool(
...
   }
+  // remove any pg_upmap_primary mapping for this pool
+  for (auto& p : osdmap.pg_upmap_primaries) {
+    if (p.first.pool() == pool) {
+      dout(10) << __func__ << " " << pool
+               << " removing obsolete pg_upmap_primaries " << p.first
+               << dendl;
+      pending_inc.old_pg_upmap_primary.insert(p.first);
+    }
+  }
```
Comment (ljflores) [inline src/mon/OSDMonitor.cc:14640]: Thanks for this @mohit84! Can you also include removal of `pending_inc.new_pg_upmap_primary`? Similar to the above section which removes `new_pg_upmap_items`.

Comment (mohit84) [reply src/mon/OSDMonitor.cc:14641]: done

**Comment by mohit84** [comment]
jenkins test make check

File: src/mon/OSDMonitor.cc (hunk 1 above)
Comment (ljflores) [reply src/mon/OSDMonitor.cc:14641]: Thank you!

**Comment by ljflores** [comment]
Will review this soon..

**Review by JoshSalomon — CHANGES_REQUESTED** [review]
Line 14639 should be replaced with a loop similar to the loop between lines 14619 and 14631

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -14629,6 +14629,16 @@ int OSDMonitor::_prepare_remove_pool(
...
     }
   }
+  // remove any pg_upmap_primary mapping for this pool
+  for (auto& p : osdmap.pg_upmap_primaries) {
+    if (p.first.pool() == pool) {
+      dout(10) << __func__ << " " << pool
+               << " removing obsolete and pending pg_upmap_primaries " << p.first
+               << dendl;
+      pending_inc.old_pg_upmap_primary.insert(p.first);
+      pending_inc.new_pg_upmap_primary.erase(p.first);
```
[the remark this answers is not in the document]
Comment (mohit84) [reply src/mon/OSDMonitor.cc:14648]: Thanks for reviewing the patch, done !!

**Comment by mohit84** [comment]
jenkins test make check

File: src/mon/OSDMonitor.cc [hunk 3]
```diff
@@ -14629,6 +14629,25 @@ int OSDMonitor::_prepare_remove_pool(
...
+  for (auto& p : osdmap.pg_upmap_primaries) {
+    if (p.first.pool() == pool) {
+      dout(10) << __func__ << " " << pool
+               << " removing obsolete pg_upmap_primaries " << p.first
+               << dendl;
+      pending_inc.old_pg_upmap_primary.insert(p.first);
+    }
+  }
+
+  // remove any new pg_upmap_primary mapping for this pool
```
Comment (JoshSalomon) [inline src/mon/OSDMonitor.cc:14642]: I would change the comment to "remove any **pending** pg_upmap_primary ..."

**Review by JoshSalomon — APPROVED** [review]
One small comment issue

File: src/mon/OSDMonitor.cc (hunk 3 above)
Comment (mohit84) [reply src/mon/OSDMonitor.cc]: done

**Review by ljflores — APPROVED** [review]
Nit: Can you change the commit title to `mon` instead of `osdmon`?

**Review by rzarzynski — APPROVED** [review]

**Comment by mohit84** [comment]
> Nit: Can you change the commit title to `mon` instead of `osdmon`?

done

**Comment by ljflores** [comment]
jenkins test windows

**Comment by ljflores** [comment]
I noticed a compilation error, so I fixed it. @mohit84 Hope you don't mind that I pushed to your branch- I would like to include this in the next squid RC, so it had to be fixed quickly.

**Comment by ljflores** [comment]
Created the squid backport to test in parallel.

**Comment by NitzanMordhai** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues67266

**Comment by mohit84** [comment]
> I noticed a compilation error, so I fixed it. @mohit84 Hope you don't mind that I pushed to your branch- I would like to include this in the next squid RC, so it had to be fixed quickly.
> 
> I verified that the fix works as expected.

Of course i don't mind it.

**Pull request merged**